### PR TITLE
Socat sidecar proxy for Couchbase container to support random ports

### PR DESCRIPTION
**Pull request by kaidowei**
inspired by #785.

- containerIsCreated hook 
- added support for couchbase 5.5
- added [ALL couchbase client ports](https://developer.couchbase.com/documentation/server/current/install/install-ports.html#page-template-attributes)
- restructured integration tests for couchbase  4.6, 5.1 and 5.5

### CODE REVIEW

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java [hunk 1]
```diff
@@ -356,6 +358,10 @@ protected void configure() {
 
     }
 
+    @SuppressWarnings({"EmptyMethod", "UnusedParameters"})
+    protected void containerIsCreated(String containerId) {
```
Comment (kiview) [inline core/src/main/java/org/testcontainers/containers/GenericContainer.java]: It's actually `containerWasCreated`, or is it still in process?

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java [hunk 2]
```diff
@@ -556,7 +562,10 @@ public void setWaitStrategy(org.testcontainers.containers.wait.strategy.WaitStra
      * @see #waitingFor(org.testcontainers.containers.wait.strategy.WaitStrategy)
      */
     protected void waitUntilContainerStarted() {
-        getWaitStrategy().waitUntilReady(this);
+        org.testcontainers.containers.wait.strategy.WaitStrategy waitStrategy = getWaitStrategy();
```
Comment (kiview) [inline core/src/main/java/org/testcontainers/containers/GenericContainer.java]: Is the import missing?

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java [hunk 3]
```diff
@@ -1028,8 +1037,8 @@ public void copyFileToContainer(MountableFile mountableLocalFile, String contain
     @Override
     public void copyFileFromContainer(String containerPath, String destinationPath) throws IOException {
 
-        if (!isRunning()) {
-            throw new IllegalStateException("copyFileToContainer can only be used while the Container is running");
+        if (!isCreated()) {
```
Comment (kiview) [inline core/src/main/java/org/testcontainers/containers/GenericContainer.java]: Kind of like above, `wasCreated()` seems more clear.
Would change the exception message accordingly to "after the container was created".

Comment (kaidowei) [reply core/src/main/java/org/testcontainers/containers/GenericContainer.java]: this is a public (default) method from `ContainerState.java` which is only used twice in the project. Still I do not feel comfortable to rename it.

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java (hunk 1 above)
Comment (kaidowei) [reply core/src/main/java/org/testcontainers/containers/GenericContainer.java]: The creation is finished. I just named it in the style of the other callbacks `containerIsStarted` and `containerIsStarting`.

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java (hunk 2 above)
Comment (kaidowei) [reply core/src/main/java/org/testcontainers/containers/GenericContainer.java]: no, this is the new `org.testcontainers.containers.wait.strategy.WaitStrategy` which is has a name clash with the deprecated `org.testcontainers.containers.wait.WaitStrategy`

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java (hunk 3 above)
Comment (kiview) [reply core/src/main/java/org/testcontainers/containers/GenericContainer.java]: Okay, let's not rename it if we break the public API.

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java (hunk 1 above)
Comment (kiview) [reply core/src/main/java/org/testcontainers/containers/GenericContainer.java]: Yes, was assuming something like this and makes sense to keep the style consistent. Just maybe less clear language.

File: modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java [hunk 4]
```diff
@@ -93,37 +112,98 @@
...
+    public List<Integer> getBoundPortNumbers() {
+        return proxy.getBoundPortNumbers();
+    }
+
+    @Override
+    public void stop() {
+        Stream.<Runnable>of(super::stop, proxy::stop, this::stopCluster).parallel().forEach(Runnable::run);
+    }
+
+    private void stopCluster() {
```
Comment (bsideup) [inline modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java]: what's the point of it? We're forcibly shutting down the container, isn't it enough?

File: modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java [hunk 5]
```diff
@@ -239,6 +331,33 @@ public void callCouchbaseRestAPI(String url, String payload) throws IOException
...
+    protected void containerIsCreated(String containerId) {
+        File tempFile = copyConfigFromContainer(containerId);
+        appendPortsToConfig(tempFile);
+        copyFileToContainer(MountableFile.forHostPath(tempFile.toPath()), STATIC_CONFIG_PATH);
+    }
+
+    private File copyConfigFromContainer(String containerId) throws IOException {
+        File tempDirectory = Files.createTempDirectory("testcontainer_" + containerId + "_").toFile();
+        tempDirectory.deleteOnExit();
+        File tempFile = new File(tempDirectory, STATIC_CONFIG_NAME);
```
Comment (bsideup) [inline modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java]: there is `Files.createTempFile()`

File: modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java [hunk 6]
```diff
@@ -239,6 +331,33 @@ public void callCouchbaseRestAPI(String url, String payload) throws IOException
...
+        tempFile.deleteOnExit();
+        logger().debug("Storing static_config in [{}].", tempFile.getAbsolutePath());
+        copyFileFromContainer(STATIC_CONFIG_LOCATION, tempFile.getPath());
+        return tempFile;
+    }
+
+    private void appendPortsToConfig(File tempFile) throws IOException {
+        for(CouchbasePort port: CouchbasePort.values()) {
+            if (! port.isBootstrap()) {
```
Comment (bsideup) [inline modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java]: code style: no space after `!`

File: modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java [hunk 7]
```diff
@@ -239,6 +331,33 @@ public void callCouchbaseRestAPI(String url, String payload) throws IOException
...
+        tempDirectory.deleteOnExit();
+        File tempFile = new File(tempDirectory, STATIC_CONFIG_NAME);
+        tempFile.deleteOnExit();
+        logger().debug("Storing static_config in [{}].", tempFile.getAbsolutePath());
+        copyFileFromContainer(STATIC_CONFIG_LOCATION, tempFile.getPath());
+        return tempFile;
+    }
+
+    private void appendPortsToConfig(File tempFile) throws IOException {
+        for(CouchbasePort port: CouchbasePort.values()) {
```
Comment (bsideup) [inline modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java]: codestyle: space before `:`

File: modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java [hunk 8]
```diff
@@ -316,4 +436,30 @@ public CouchbaseContainer withGamesIMSample(boolean gamesIMSample) {
...
+        FTS_SSL("fts_ssl_port", 18094, false),
+        CBAS_SSL("cbas_ssl_port", 18095, false),
+        EVENTING_SSL("eventing_ssl_port", 18096, false),
+        ;
+
+        final String name;
+
+        final int originalPort;
+
+        final boolean isBootstrap;
```
Comment (bsideup) [inline modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java]: I would rename to `dynamic`, and also mark MEMCACHED as dynamic (as it was in the previous implementation)

Comment (kaidowei) [reply modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java]: renaming is fine.


Problem with MEMCACHED is, that when marking them as dynamic, the sdk does not work:
```
11:26:49.578 WARN  com.couchbase.client.core.endpoint.Endpoint - [localhost:11210][KeyValueEndpoint]: Could not connect to endpoint on reconnect attempt 5, retrying with delay 32 MILLISECONDS: com.couchbase.client.deps.io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:11210
11:26:49.612 WARN  com.couchbase.client.core.endpoint.Endpoint - [localhost:11210][KeyValueEndpoint]: Could not connect to remote socket: Connection refused: localhost/127.0.0.1:11210
```
It uses the right mapped port for bootstrapping but then **discovers** the unmapped port from the cluster and then tries to connect to that port and fails with connection refused.

**Pull request merged**